File: amr/mapping/occupancy_grid_mapper.py
"""Log-odds occupancy grid mapper.

Update rule per plan §Task 1.2:
- For each beam in scan[::beam_subsample]:
  * Compute the endpoint in world coords via pose ⊕ (r·cosθ, r·sinθ, 0)
  * Run Bresenham from robot cell to endpoint cell
  * All cells on the ray (except the last) get log_odds += l_free
  * The endpoint cell gets log_odds += l_occ  only if the beam is a valid return
  * For no-return beams: trace free to range_max * 0.99, skip endpoint update
  * Clamp log_odds to ±l_clamp

to_occupancy_grid thresholds (plan §2 item 5):
  p = 1 - 1/(1 + exp(log_odds))
  p > occupied_thresh → 100
  p < free_thresh     → 0
  else                → -1  (unknown, or unobserved cells where log_odds == 0)
"""
from __future__ import annotations

import math
from typing import Tuple

import numpy as np

from amr.core.config import MappingConfig
from amr.core.geometry import bresenham
from amr.core.types import LaserScan, OccupancyGrid, Pose2D


class OccupancyGridMapper:
    """Log-odds occupancy grid mapper."""

    log_odds: np.ndarray  # float32 (rows, cols), 0 = unobserved

    def __init__(self, cfg: MappingConfig,
                 size_m: Tuple[float, float],
                 origin_xy: Tuple[float, float] = (0.0, 0.0)) -> None:
        self._cfg = cfg
        self._resolution = cfg.resolution
        self._origin_x = float(origin_xy[0])
        self._origin_y = float(origin_xy[1])

        cols = int(round(size_m[0] / cfg.resolution))
        rows = int(round(size_m[1] / cfg.resolution))
        self.log_odds = np.zeros((rows, cols), dtype=np.float32)

    # ------------------------------------------------------------------
    # Private helpers
    # ------------------------------------------------------------------

    def _world_to_cell(self, x: float, y: float) -> Tuple[int, int]:
        row = int(math.floor((y - self._origin_y) / self._resolution))
        col = int(math.floor((x - self._origin_x) / self._resolution))
        return row, col

    def _in_bounds(self, row: int, col: int) -> bool:
        return 0 <= row < self.log_odds.shape[0] and 0 <= col < self.log_odds.shape[1]

    # ------------------------------------------------------------------
    # Public interface
    # ------------------------------------------------------------------

    def update(self, pose: Pose2D, scan: LaserScan) -> None:
        """Integrate one laser scan at the given pose into the log-odds map."""
        cfg = self._cfg
        robot_r, robot_c = self._world_to_cell(pose.x, pose.y)

        angles = scan.angles()
        valid = scan.valid_mask()

        # Subsample beams
        indices = range(0, scan.num_beams, cfg.beam_subsample)

        cos_th = math.cos(pose.theta)
        sin_th = math.sin(pose.theta)

        for i in indices:
            angle_b = angles[i]
            is_valid = bool(valid[i])

            if is_valid:
                r_dist = float(scan.ranges[i])
            else:
                # Trace free space up to range_max * 0.99
                r_dist = scan.range_max * 0.99

            # Beam direction in world frame
            beam_cos = math.cos(pose.theta + angle_b)
            beam_sin = math.sin(pose.theta + angle_b)

            # Endpoint in world coords
            end_x = pose.x + r_dist * beam_cos
            end_y = pose.y + r_dist * beam_sin
            end_r, end_c = self._world_to_cell(end_x, end_y)

            # Bresenham ray from robot cell to endpoint cell
            cells = bresenham(robot_r, robot_c, end_r, end_c)

            # Mark all cells except the last as free
            for cell_r, cell_c in cells[:-1]:
                if self._in_bounds(cell_r, cell_c):
                    self.log_odds[cell_r, cell_c] = np.clip(
                        self.log_odds[cell_r, cell_c] + cfg.l_free,
                        -cfg.l_clamp, cfg.l_clamp
                    )

            # Mark the endpoint cell as occupied (only for valid returns)
            if is_valid and self._in_bounds(end_r, end_c):
                self.log_odds[end_r, end_c] = np.clip(
                    self.log_odds[end_r, end_c] + cfg.l_occ,
                    -cfg.l_clamp, cfg.l_clamp
                )

    def to_occupancy_grid(self) -> OccupancyGrid:
        """Convert log-odds map to ROS-style OccupancyGrid.

        p = 1 - 1/(1 + exp(log_odds))
        p > occupied_thresh → 100
        p < free_thresh     → 0
        else                → -1
        """
        cfg = self._cfg
        # p_occ = 1 - 1/(1+exp(log_odds)) = exp(log_odds)/(1+exp(log_odds))
        # Equivalent: p = sigmoid(log_odds)
        # p = 0.5 when log_odds = 0 (unobserved)
        lo = self.log_odds.astype(np.float64)
        p = 1.0 - 1.0 / (1.0 + np.exp(lo))

        data = np.full(self.log_odds.shape, -1, dtype=np.int8)
        data[p > cfg.occupied_thresh] = 100
        data[p < cfg.free_thresh] = 0
        # Cells where log_odds == 0 exactly → p == 0.5 → unknown (-1)

        return OccupancyGrid(
            resolution=self._resolution,
            origin_x=self._origin_x,
            origin_y=self._origin_y,
            data=data,
        )

File: src/amr_mapping/src/occupancy_grid_mapper.cpp
// SPDX-License-Identifier: Apache-2.0
// Log-odds occupancy-grid mapper. Faithful C++ port of
// amr/mapping/occupancy_grid_mapper.py.
#include "amr_mapping/occupancy_grid_mapper.hpp"

#include <algorithm>
#include <cmath>
#include <utility>

#include "amr_core/geometry.hpp"

namespace amr_mapping {

OccupancyGridMapper::OccupancyGridMapper(const amr_core::MappingConfig& cfg,
                                         std::pair<double, double> size_m,
                                         std::pair<double, double> origin_xy)
    : cfg_(cfg),
      resolution_(cfg.resolution),
      origin_x_(origin_xy.first),
      origin_y_(origin_xy.second) {
  cols_ = static_cast<int>(std::lround(size_m.first / cfg.resolution));
  rows_ = static_cast<int>(std::lround(size_m.second / cfg.resolution));
  log_odds_.assign(static_cast<std::size_t>(rows_) * cols_, 0.0f);
}

void OccupancyGridMapper::world_to_cell(double x, double y, int& row,
                                        int& col) const {
  row = static_cast<int>(std::floor((y - origin_y_) / resolution_));
  col = static_cast<int>(std::floor((x - origin_x_) / resolution_));
}

void OccupancyGridMapper::update(const amr_core::Pose2D& pose,
                                 const amr_core::LaserScan& scan) {
  int robot_r = 0;
  int robot_c = 0;
  world_to_cell(pose.x, pose.y, robot_r, robot_c);

  const std::vector<double> angles = scan.angles();
  const std::vector<char> valid = scan.valid_mask();
  const int num_beams = scan.num_beams();

  const float l_free = static_cast<float>(cfg_.l_free);
  const float l_occ = static_cast<float>(cfg_.l_occ);
  const float l_clamp = static_cast<float>(cfg_.l_clamp);

  for (int i = 0; i < num_beams; i += cfg_.beam_subsample) {
    const double angle_b = angles[static_cast<std::size_t>(i)];
    const bool is_valid = valid[static_cast<std::size_t>(i)] != 0;

    double r_dist;
    if (is_valid) {
      r_dist = scan.ranges[static_cast<std::size_t>(i)];
    } else {
      // No-return beam: trace free space up to range_max * 0.99.
      r_dist = scan.range_max * 0.99;
    }

    // Beam direction in the world frame.
    const double beam_cos = std::cos(pose.theta + angle_b);
    const double beam_sin = std::sin(pose.theta + angle_b);

    const double end_x = pose.x + r_dist * beam_cos;
    const double end_y = pose.y + r_dist * beam_sin;
    int end_r = 0;
    int end_c = 0;
    world_to_cell(end_x, end_y, end_r, end_c);

    // Bresenham ray from robot cell to endpoint cell ((row, col) pairs).
    const std::vector<std::pair<int, int>> cells =
        amr_core::bresenham(robot_r, robot_c, end_r, end_c);

    // Mark every cell except the last as free.
    if (cells.size() > 1) {
      for (std::size_t k = 0; k + 1 < cells.size(); ++k) {
        const int cr = cells[k].first;
        const int cc = cells[k].second;
        if (in_bounds(cr, cc)) {
          float& lo = log_odds_[static_cast<std::size_t>(cr) * cols_ + cc];
          lo = std::min(std::max(lo + l_free, -l_clamp), l_clamp);
        }
      }
    }

    // Mark the endpoint cell as occupied (only for valid returns).
    if (is_valid && in_bounds(end_r, end_c)) {
      float& lo = log_odds_[static_cast<std::size_t>(end_r) * cols_ + end_c];
      lo = std::min(std::max(lo + l_occ, -l_clamp), l_clamp);
    }
  }
}

amr_core::OccupancyGrid OccupancyGridMapper::to_occupancy_grid() const {
  amr_core::OccupancyGrid grid;
  grid.resolution = resolution_;
  grid.origin_x = origin_x_;
  grid.origin_y = origin_y_;
  grid.rows = rows_;
  grid.cols = cols_;
  grid.data.assign(log_odds_.size(), static_cast<int8_t>(-1));

  for (std::size_t i = 0; i < log_odds_.size(); ++i) {
    const double lo = static_cast<double>(log_odds_[i]);
    const double p = 1.0 - 1.0 / (1.0 + std::exp(lo));
    if (p > cfg_.occupied_thresh) {
      grid.data[i] = 100;
    } else if (p < cfg_.free_thresh) {
      grid.data[i] = 0;
    }
    // else stays -1 (unknown; log_odds == 0 -> p == 0.5).
  }
  return grid;
}

}  // namespace amr_mapping
